### PR TITLE
Websocket interface

### LINKED ISSUE #16: Create Ruby Client for Websockets

### PR DESCRIPTION
**Pull request by ryanmelt**
closes #16

### CODE REVIEW

File: openc3/lib/openc3/script/web_socket_api.rb [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+  # Timeline WebSocket
+  class TimelineEventsWebSocketApi < CmdTlmWebSocketApi
+    def initialize(history_count: 0, url: nil, write_timeout: 10.0, read_timeout: 10.0, connect_timeout: 5.0, authentication: nil, scope: $openc3_scope)
+      @identifier = {
+        channel: "TimelineEventsChannel",
+        history_count: history_count
+      }
+      super(url: url, write_timeout: write_timeout, read_timeout: read_timeout, connect_timeout: connect_timeout, authentication: authentication, scope: scope)
+    end
+  end
```
Comment (jmthomas) [inline openc3/lib/openc3/script/web_socket_api.rb:298]: All these classes are a lot of boilerplate. Seems like the only thing they're doing is setting the @identifier channel to the basically the same name as the class. Not sure if a Class.new would help but maybe it's more trouble that it's worth.

File: openc3/lib/openc3/streams/web_socket_client_stream.rb [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# This file may also be used under the terms of a commercial license
+# if purchased from OpenC3, Inc.
+
+require 'openssl'
+require 'openc3/streams/tcpip_client_stream'
+require 'websocket'
```
Comment (jmthomas) [inline openc3/lib/openc3/streams/web_socket_client_stream.rb:21]: [websocket-ruby ](https://github.com/imanel/websocket-ruby)

File: openc3-cosmos-script-runner-api/app/channels/application_cable/connection.rb [hunk 3]
```diff
@@ -12,15 +12,36 @@
...
+      self.uuid = SecureRandom.uuid
+      self.scope = request.headers['HTTP_X_OPENC3_SCOPE'] || request.query_parameters[:scope]
+    end
+
+    def authorization(permission)
+      begin
+        authorize(
+          permission: permission,
+          scope: request.headers['HTTP_X_OPENC3_SCOPE'] || request.query_parameters[:scope],
+          token: request.headers['HTTP_AUTHORIZATION'] || request.query_parameters[:authorization],
```
Comment (jmthomas) [inline openc3-cosmos-script-runner-api/app/channels/application_cable/connection.rb:39]: Who sets these headers? Or are they always in the query_parameters?

File: openc3-cosmos-cmd-tlm-api/app/channels/application_cable/connection.rb [hunk 4]
```diff
@@ -12,25 +12,36 @@
...
-      #   request.headers["X-API-TOKEN"]
+      self.scope = request.headers['HTTP_X_OPENC3_SCOPE'] || request.query_parameters[:scope]
+    end
+
+    def authorization(permission)
+      begin
+        authorize(
+          permission: permission,
+          scope: request.headers['HTTP_X_OPENC3_SCOPE'] || request.query_parameters[:scope],
+          token: request.headers['HTTP_AUTHORIZATION'] || request.query_parameters[:authorization],
```
Comment (jmthomas) [inline openc3-cosmos-cmd-tlm-api/app/channels/application_cable/connection.rb:39]: When are they in the header vs the request_parameters?

**Review by jmthomas — COMMENTED** [review]
This is quite magical. I'm glad you're a wizard!

File: openc3-cosmos-cmd-tlm-api/app/channels/application_cable/connection.rb (hunk 4 above)
Comment (ryanmelt) [reply openc3-cosmos-cmd-tlm-api/app/channels/application_cable/connection.rb:39]: Unfortunately you can't use headers from webpage javascript.   I was using that from the Ruby code but then just standardized on using the query parameters everywhere.  I can remove the headers option.

File: openc3-cosmos-init/plugins/packages/openc3-tool-common/src/services/cable.js [hunk 5]
```diff
@@ -26,18 +26,26 @@ ActionCable.ConnectionMonitor.staleThreshold = 10
...
     return OpenC3Auth.updateToken(OpenC3Auth.defaultMinValidity).then(() => {
+      if (this._cable == null) {
+        let final_url =
+          this._url +
+          '?scope=' +
+          window.openc3Scope +
+          '&authorization=' +
+          localStorage.openc3Token
+        this._cable = ActionCable.createConsumer(final_url)
+      }
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3-cosmos-init/plugins/packages/openc3-tool-common/src/services/cable.js:45]: Yes, from all the subscriptions.  Scope and token should still be used in actions.

File: openc3-cosmos-script-runner-api/app/channels/application_cable/connection.rb (hunk 3 above)
Comment (ryanmelt) [reply openc3-cosmos-script-runner-api/app/channels/application_cable/connection.rb:39]: Always query parameters.  Will update.

File: openc3/lib/openc3/streams/web_socket_client_stream.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3/lib/openc3/streams/web_socket_client_stream.rb:21]: correct.

File: openc3/lib/openc3/script/web_socket_api.rb [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+              next
+            end
+          end
+          if Hash === json_hash['message']
+            return json_hash['message']
+          else
+            return JSON.parse(json_hash['message'], allow_nan: true, create_additions: true)
+          end
+        end
+        return message
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/script/web_socket_api.rb:89]: Yes

File: openc3/lib/openc3/script/web_socket_api.rb [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+    end
+
+    # Read the next message with json parsing, filtering, and timeout support
+    def read(ignore_keep_alive: true, timeout: nil)
+      start_time = Time.now
+      while true
+        message = read_message()
+        if message
+          json_hash = JSON.parse(message, allow_nan: true, create_additions: true)
+          if ignore_keep_alive
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/script/web_socket_api.rb:60]: Ignore protocol messages would be a better name.

File: openc3/lib/openc3/script/web_socket_api.rb [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+            type = json_hash['type']
+            if type # ping, welcome, confirm_subscription, reject_subscription, disconnect
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/script/web_socket_api.rb:62]: Nope.

File: openc3/lib/openc3/script/web_socket_api.rb [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+              if type == 'disconnect'
+                if json_hash['reason'] == 'unauthorized'
+                  raise "Unauthorized"
+                end
+              end
+              if type == 'reject_subscription'
+                raise "Subscription Rejected"
+              end
+              if timeout
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/script/web_socket_api.rb:71]: Usually protocol messages like ping.

File: openc3/lib/openc3/script/web_socket_api.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3/lib/openc3/script/web_socket_api.rb:298]: They are very similar but at only like 6 lines I think good enough.

**Pull request merged**